[RFC] eval: handle profiling and verbose in system()

maurelio1234 opened the pull request:
This is my take on issue #8362.

The code to handle verbose and profiling seems to have been lost on a
refactoring of `evalbuf.c` (in vim codebase) into `eval.c` (neovim).

Vim's code calls `call_shell` directly from `get_system_output_as_rettv`()
whereas in neovim this function has been rewritten to not call `call_shell` but to call
`os_system` via `do_os_system`, loosing the support for profiling and verbose.

Changing the code to call `call_shell` from `get_system_output_as_rettv`
seems to be too complicated to be worth it on the current version of the
code. It seems to be simpler to just re-add the code to handling the
missing bits on its current version.

Current TODO list:

- [x] Remove empty printed line
- [x] Write test for verbose
- [x] Write test for profiling
- [x] This code doesn't print exactly the same thing as Vim... but it is coherent to other Nvim calls... is the difference important? 
- [x] os_system wrapper
- [x] keep this list in alphabetical order. same for the other one.
- [x] fix integration test

Commit:
eval: handle profiling and verbose in system()

The code to handle verbose and profiling seems to have been lost on a
refactoring of `evalbuf.c` (in vim codebase) into `eval.c` (neovim).

Vim's code calls `call_shell` directly from `get_system_output_as_rettv`()
whereas in neovim this function has been rewritten to not call `call_shell` but to call
`os_system` via `do_os_system`, loosing the support for profiling and verbose.

Changing the code to call `call_shell` from `get_system_output_as_rettv`
seems to be too complicated to be worth it on the current version of the
code. It seems to be simpler to just re-add the code to handling the
missing bits on its current version.

justinmk commented:
Seems like a fine plan. 

>  is the difference important?

No, it's ok. In fact, another behavior change I think we should do (maybe different PR): don't show a hit-enter prompt if nothing was returned from `:!` or `:echo system()`. E.g., this shouldn't show hit-enter:

    :!printf ''

Commit:
test: functional

Commit:
test: remove extra feed command

maurelio1234 commented on src/nvim/eval.c:
```diff
@@ -16406,11 +16409,28 @@ static void get_system_output_as_rettv(typval_T *argvars, typval_T *rettv,
+    char buf[NUMBUFLEN];
+    verbose_enter();
+    smsg(_("Calling shell to execute: \"%s\""), tv_get_string_buf(argvars, buf));
+    smsg(_("")); // TODO: why do I need to add this?
```
For some reason... it doesn't print the first message... but I don't know the code base enough to see if something funny is happening. I hope I have time to investigate that during the weekend.

The result of the functional test without L16416:

```test/functional/ui/screen.lua:307: Row 10 did not match.
Expected:
  |~                                                    |
  |~                                                    |
  |~                                                    |
  |~                                                    |
  |~                                                    |
  |~                                                    |
  |~                                                    |
  |~                                                    |
  |~                                                    |
  |*Calling shell to execute: "echo hi"                  |
  |                                                     |
  |hi                                                   |
  |                                                     |
  |Press ENTER or type command to continue^              |
Actual:
  |~                                                    |
  |~                                                    |
  |~                                                    |
  |~                                                    |
  |~                                                    |
  |~                                                    |
  |~                                                    |
  |~                                                    |
  |~                                                    |
  |*                                                     |
  |                                                     |
  |hi                                                   |
  |                                                     |
  |Press ENTER or type command to continue^              |

To print the expect() call that would assert the current screen state, use
screen:snapshot_util(). In case of non-deterministic failures, use
screen:redraw_debug() to show all intermediate screen states.  
```

justinmk replied:
If I do:

    git grep smsg

I see that it's a common pattern, e.g.:

```c
    smsg(_("Exception caught: %s"), excp->value);
    msg_puts("\n");  // don't overwrite this either
```

So I think it's fine as you have it. Though perhaps change it to `msg_puts("\n"); `.

Commit:
test: test profiling

Commit:
eval: remove extra call to smsg

I still do not know the code base enough to know why the way it was
written before didn't work.

In this commit I changed it to mirror the
way other functions in the same file print verbose information (calling
`verbose_enter_scroll` / `verbose_leave_scroll`) and it worked as expected.

I added a msg_puts to separate the error message from the the
command output. I added a double '\n' just to make it match the current
output for the '!' counterpart (i.e. test `:4verb !echo hi`).

This commit also changes some white spaces in the functional test to
match the actual output of the call.

Commit:
test: exposing near() assertion

Commit:
test: rewrite functional test to make it more readable

Commit:
test: extracting self time using lua code instead of eval'ed vim code

Commit:
lint: split line to make it less than 80 chars

maurelio1234 commented:
Some builds seem to fail for reasons unrelated to my code... is that normal?

Commit:
test: make selftime local

justinmk commented:
> Some builds seem to fail for reasons unrelated to my code... is that normal?

Yes, unfortunately. https://github.com/neovim/neovim/issues/8102 is the main problem.

justinmk commented on test/helpers.lua:
```diff
@@ -696,6 +699,7 @@ local module = {
   mergedicts_copy = mergedicts_copy,
   neq = neq,
   ok = ok,
+  near = near,
```
keep this list in alphabetical order

justinmk commented on test/functional/eval/system_spec.lua:
```diff
@@ -203,6 +203,45 @@ describe('system()', function()
+      local tempfile = helpers.tmpname()
+
+      feed(':function! AlmostNoSelfTime()<cr>')
+      feed('echo system("sleep 2")<cr>')
```
This just fails on Windows (no `sleep` command), so the comparison of `near(selftime, 0, 0.01)` is fragile.

I think it's enough to just check that `:profile` "did something", don't need to compare times. Though the effort is appreciated, but flakey tests are something we need to avoid (more) :)

justinmk commented on src/nvim/eval.c:
```diff
@@ -16377,9 +16377,12 @@ static list_T *string_to_list(const char *str, size_t len, const bool keepempty)
   return list;
 }
 
+// os_call_shell wrapper. Handles 'verbose', :profile, and v:shell_error.
```
`os_system` wrapper

Commit:
style: keep list in alphabetic order

Commit:
comment: fix referred function

Commit:
test: check if self and total time are reported

Instead of sleeping, which makes test not work in windows, and unstable
by nature.

Commit:
style: keep declarations in alphabetical order

maurelio1234 commented:
Hi, is there something missing here?

maurelio1234 commented:
> No, it's ok. In fact, another behavior change I think we should do (maybe different PR): don't show a hit-enter prompt if nothing was returned from :! or :echo system(). E.g., this shouldn't show hit-enter:

@justinmk concerning this comment, I just tested this in Vim 8.1 and it behaves exactly like neovim, it shows a hit enter prompt even if the called commend returns nothing.

justinmk commented:
> I just tested this in Vim 8.1 and it behaves exactly like neovim, it shows a hit enter prompt even if the called commend returns nothing.

Yes, but it's not useful. So I think Nvim should diverge there (not in this PR though).
